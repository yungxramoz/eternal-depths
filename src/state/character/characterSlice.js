import { createSlice } from '@reduxjs/toolkit'
import Character from '../../models/Character'

const initialState = {
  current: new Character(),
  availableAttributePoints: 2,
}

export const characterSlice = createSlice({
  name: 'character',
  initialState,
  reducers: {
    setName: (state, { payload }) => {
      state.current.name = payload
    },
    setLook: (state, { payload }) => {
      state.current.look = payload
    },
    assignAttributePoint: (
      state,
      { payload: { strength, health, agility, precision } },
    ) => {
      const total = strength + health + agility + precision
      if (state.availableAttributePoints >= total) {
        state.current.strength += strength
        state.current.health += health
        state.current.agility += agility
        state.current.precision += precision
        state.availableAttributePoints -= total
      }
    },
  },
})

export const { setName, setLook, assignAttributePoint } = characterSlice.actions

export default characterSlice.reducer
