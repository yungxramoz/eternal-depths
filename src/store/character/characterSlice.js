import { createSelector, createSlice } from '@reduxjs/toolkit'
import { ATTACK, BASE_ATTACK } from '../../constants/attack-type'

const initialState = {
  current: {
    look: 1,
    name: '',
    level: 1,
    xp: 0,
    stats: {
      health: 1,
      strength: 1,
      agility: 1,
      precision: 1,
    },
    items: {
      helmet: null,
      armor: null,
      weapon: {
        name: 'Rusty Sword',
        stats: {
          minDamage: 1,
          maxDamage: 3,
        },
      },
      shield: null,
      greaves: null,
    },
    attacks: [
      {
        ...BASE_ATTACK,
        id: 1,
        currentCooldown: 0,
      },
      {
        ...ATTACK.DEATH_BLOW,
        id: 2,
        currentCooldown: 0,
      },
    ],
  },
  availableAttributePoints: 2,
}

export const characterSlice = createSlice({
  name: 'character',
  initialState,
  selectors: {
    calculatedCharacterStats: createSelector(
      (state) => state.current,
      (current) => {
        const stats = {
          health: current.stats.health,
          strength: current.stats.strength,
          agility: current.stats.agility,
          precision: current.stats.precision,
        }
        const items = [
          current.items.helmet,
          current.items.armor,
          current.items.shield,
          current.items.greaves,
        ]
        for (const item of items) {
          if (item) {
            for (const stat in item.stats) {
              stats[stat] += item.stats[stat]
            }
          }
        }
        return stats
      },
    ),
  },
  reducers: {
    setName: (state, { payload }) => {
      state.current.name = payload
    },
    setLook: (state, { payload }) => {
      state.current.look = payload
    },
    assignAttributePoint: (
      state,
      { payload: { health = 0, strength = 0, agility = 0, precision = 0 } },
    ) => {
      const total = health + strength + agility + precision
      if (state.availableAttributePoints >= total) {
        state.current.stats.health += health
        state.current.stats.strength += strength
        state.current.stats.agility += agility
        state.current.stats.precision += precision
        state.availableAttributePoints -= total
        characterSlice.caseReducers.setHpToMax(state)
      }
    },
    setHpToMax: (state) => {
      const maxHp =
        characterSlice.getSelectors().calculatedCharacterStats(state).health *
          10 +
        90
      state.current.maxHp = maxHp
      state.current.hp = maxHp
    },
    recalculateMaxHp: (state) => {
      const maxHp =
        characterSlice.getSelectors().calculatedCharacterStats(state).health *
          10 +
        90
      state.current.maxHp = maxHp
    },
    recoverHp: (state, { payload }) => {
      state.current.hp += payload
      if (state.current.hp > state.current.maxHp) {
        state.current.hp = state.current.maxHp
      }
    },
    equipItem: (state, { payload }) => {
      state.current.items[payload.type] = payload
      characterSlice.caseReducers.recalculateMaxHp(state)
    },
    attackEffects: (state, { payload: { attack, dealtDamage } }) => {
      if (attack.selfHealAmount != null) {
        let heal = 0
        if (attack.selfHealAmount === 'auto') {
          heal = dealtDamage
        } else {
          heal = attack.selfHealAmount
        }
        console.log('heal', heal)
        state.current.hp += heal
        if (state.current.hp > state.current.maxHp) {
          state.current.hp = state.current.maxHp
        }
      }
      if (attack.selfInflictedAmount > 0) {
        state.current.hp -= attack.selfInflictedAmount
        if (state.current.hp <= 0) {
          state.current.hp = 0
        }
        console.log('selfInflictedAmount', attack.selfInflictedAmount)
      }

      state.current.attacks.forEach((a) => {
        if (a.currentCooldown > 0) {
          a.currentCooldown -= 1
        }
      })

      state.current.attacks.find((a) => a.id === attack.id).currentCooldown =
        attack.cooldown
      console.log('attack.cooldown', attack.cooldown)
    },
    damageCharacter: (state, { payload }) => {
      state.current.hp -= payload
      if (state.current.hp <= 0) {
        state.current.hp = 0
      }
    },
  },
})

export const { calculatedCharacterStats } = characterSlice.selectors
export const {
  setName,
  setLook,
  assignAttributePoint,
  setHpToMax,
  recoverHp,
  equipItem,
  damageCharacter,
  attackEffects,
} = characterSlice.actions

export default characterSlice.reducer
