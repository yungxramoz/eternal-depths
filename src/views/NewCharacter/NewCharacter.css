.btn-container {
  display: flex;
  justify-content: center;
  margin-top: 1rem;
}

.look-container {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  margin: 1rem 0;
}
