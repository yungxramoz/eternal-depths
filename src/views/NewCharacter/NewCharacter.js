import React from 'react'
import RpgButton from '../../components/atoms/RpgButton/RpgButton'
import RpgContainer from '../../components/templates/RpgContainer/RpgContainer'
import RpgSeparator from '../../components/templates/RpgSeperator/RpgSeparator'
import './NewCharacter.css'
import IconButton from '../../components/atoms/IconButton/IconButton'
import { useDispatch, useSelector } from 'react-redux'
import { assignAttributePoint } from '../../state/character/characterSlice'

const NewCharacter = () => {
  const character = useSelector((state) => state.character.current)
  const availableAttributePoints = useSelector((state) => state.character.availableAttributePoints)

  const dispatch = useDispatch()

  return (
    <RpgContainer fullPage>
      <h1>New Character</h1>
      <RpgSeparator />
      <div className="look-container">
        <IconButton iconName="chevron-left" />
        <div className="rpgui-icon helmet-slot" />
        <IconButton iconName="chevron-right" />
      </div>
      <input type="text" placeholder="Name" value={character.name} />
      <RpgSeparator />
      <RpgContainer golden className="attribute-container">
        <h2>Attributes</h2>
        <p>Available points: {availableAttributePoints}</p>
        <RpgSeparator golden={false} />
        <div className="attribute">
          <p>Health</p>
          <p>- {character.health} +</p>
        </div>
        <div className="attribute">
          <p>Strength</p>
          <p>- {character.strength}</p>
          <IconButton iconName="chevron-right" onClick={() => dispatch(assignAttributePoint('strength'))} />
        </div>
        <div className="attribute">
          <p>Agility</p>
          <p>- {character.agility} +</p>
        </div>
        <div className="attribute">
          <p>Precision</p>
          <p>- {character.precision} +</p>
        </div>
      </RpgContainer>
      <div className='btn-container'>
        <RpgButton text="Create" />
      </div>
    </RpgContainer>
  )
}

export default NewCharacter
