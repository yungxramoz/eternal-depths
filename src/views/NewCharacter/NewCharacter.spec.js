import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { MemoryRouter as Router } from 'react-router-dom'
import NewCharacter from './NewCharacter'
import { configureStore } from '@reduxjs/toolkit'
import characterSlice from '../../store/character/characterSlice'

jest.mock(
  '../../components/organisms/CharacterChooser/CharacterChooser',
  () => {
    return ({ onCharacterChange }) => (
      <button onClick={() => onCharacterChange(1)}>Change character</button>
    )
  },
)

const preloadedState = {
  character: {
    current: {
      characterLook: 1,
      name: '',
      stats: {
        health: 1,
        strength: 1,
        agility: 1,
        precision: 1,
      },
    },
    availableAttributePoints: 2,
  },
}

describe('NewCharacter', () => {
  let store
  beforeEach(() => {
    store = configureStore({
      reducer: {
        character: characterSlice,
      },
      preloadedState,
    })
  })
  it('renders correctly', () => {
    render(
      <Router>
        <Provider store={store}>
          <NewCharacter />
        </Provider>
      </Router>,
    )
    expect(document.body).toMatchSnapshot('NewCharacter')
  })
  it('updates localName when the input changes', () => {
    render(
      <Router>
        <Provider store={store}>
          <NewCharacter />
        </Provider>
      </Router>,
    )
    fireEvent.change(screen.getByPlaceholderText('Name'), {
      target: { value: 'Test Name' },
    })
    expect(screen.getByPlaceholderText('Name').value).toBe('Test Name')
  })

  it('updates characterLook when CharacterChooser changes', () => {
    render(
      <Router>
        <Provider store={store}>
          <NewCharacter />
        </Provider>
      </Router>,
    )
    fireEvent.click(screen.getByText('Change character'))
  })

  it('dispatches actions when "Create" button is clicked', () => {
    store = configureStore({
      reducer: {
        character: characterSlice,
      },
      preloadedState: {
        character: {
          ...preloadedState.character,
          availableAttributePoints: 0,
        },
      },
    })
    store.dispatch = jest.fn()
    render(
      <Router>
        <Provider store={store}>
          <NewCharacter />
        </Provider>
      </Router>,
    )
    fireEvent.change(screen.getByPlaceholderText('Name'), {
      target: { value: 'Test Name' },
    })
    fireEvent.click(screen.getByText('Change character'))

    fireEvent.click(screen.getByText('Create'))

    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'character/setName' }),
    )
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'character/setLook' }),
    )
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'character/assignAttributePoint' }),
    )
    expect(store.dispatch).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/gameStart' }),
    )
  })
})
