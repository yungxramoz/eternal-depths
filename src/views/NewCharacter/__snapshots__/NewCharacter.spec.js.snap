// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewCharacter renders correctly: NewCharacter 1`] = `
<body>
  <div>
    <div
      class="rpgui-container container framed full-page scrollable"
    >
      <button
        class="icon-button large align-self-start"
      >
        <svg
          fill="none"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 11v2H8v2H6v-2H4v-2h2V9h2v2h12zM10 7H8v2h2V7zm0 0h2V5h-2v2zm0 10H8v-2h2v2zm0 0h2v2h-2v-2z"
            fill="currentColor"
          />
        </svg>
      </button>
      <h1>
        New Character
      </h1>
      <input
        placeholder="Name"
        type="text"
        value=""
      />
      <hr
        class="separator golden"
      />
      <div
        class="rpgui-container container framed framed-golden attribute-container"
      >
        <h2>
          Attributes
        </h2>
        <p>
          Available points: 
          2
        </p>
        <hr
          class="separator"
        />
        <div
          class="attribute"
        >
          <p>
            Health
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="attribute"
        >
          <p>
            Strength
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="attribute"
        >
          <p>
            Agility
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="attribute"
        >
          <p>
            Precision
          </p>
          <div
            class="attr-controller"
          >
            <button
              class="icon-button medium "
              disabled=""
            >
              <svg
                fill="currentColor"
                id="Minus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="2"
                  width="22"
                  x="1"
                  y="11"
                />
              </svg>
            </button>
            <p>
              1
            </p>
            <button
              class="icon-button medium "
            >
              <svg
                fill="currentColor"
                id="Plus"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="23 11 23 13 13 13 13 23 11 23 11 13 1 13 1 11 11 11 11 1 13 1 13 11 23 11"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="btn-container"
      >
        <button
          class="rpgui-button btn"
          disabled=""
        >
          Create
        </button>
      </div>
    </div>
  </div>
</body>
`;
